File: C#/C#.NET/modul_2_OOP/L02Datenkapselung/Exercise1-ViewDrivenHamster/Hamster.cs
namespace Hamster;

public class Hamster
{
    // Felder

    // Eigenschaften (Properties)
    public (int x, int y) Position { get; set; }
    public string Representation { get; private set; }
    public static string HungryRepresentation { get; } = "😡";
    public static string FedRepresentation { get; } = "🐹";
    public string RememberSymbolOnPlane { get; set; } // Problematisch! Wir sehen bald warum...

    public bool IsHungry { get; private set; }


    // Beziehungen
    private Plane _plane;
    private List<Seedling> _mouth = new List<Seedling>();

    // Konstruktor
    public Hamster(Plane plane)
    {
        IsHungry = false;
        Representation = FedRepresentation;
        _plane = plane;

        RememberSymbolOnPlane = Plane.EarthRepresentation; // Problematisch! Wir sehen bald warum...

        PositionAndManageHamster();
    }

    // Methoden
    private void PositionAndManageHamster()
    {
        var random = new Random();
        bool done;
        int x, y;

        do
        {
            x = random.Next(_plane.Size);
            y = random.Next(_plane.Size);

            done = _plane.AssignInitialPosition(this, (x, y));
        } while (!done);

        Position = (x, y);
    }

    public void Move()
    {
        var random = new Random();
        int index = random.Next(Enum.GetValues<Direction>().Length);
        var direction = Enum.GetValues<Direction>()[index];

        _plane.Position(this, direction);
    }

    public void NutritionBehaviour()
    {
        var random = new Random();

        // Zufällig hungrig werden
        if (random.NextDouble() < 0.1)
        {
            IsHungry = true;
            Representation = HungryRepresentation;
        }

        // Sehr Problematisch! Die Position in dem Dictionary und das merken des Seedlingling Symbols ist nicht konsistent,
        // falls wir diese nicht korrekt überschreiben, wenn wir einen Seedlingling essen!
        // Es muss also in HamsterIsStoringSeedlings und HamsterIsEatingSeedlings der Code hamster.RememberSymbolOnPlane = EarthRepresentation; aufgerufen werden.
        // Falls wir das vergessen, merken wir uns auch Hamster Symbole wenn zwei hamster gleichzeitg auf ein Feld gehen!
        // Siehe Klasse Plane für weitere Erklärungen.

        if (RememberSymbolOnPlane == Seedling.Representation) // view logic: bad
        //if (_plane.ContainsSeedling(Position)) // state logic: good
        { 
            if (IsHungry)
            {
                EatSeedlingFromTile();
            }
            else
            {
                StoreInMouth();
            }
        }
        else
        {
            if (IsHungry && _mouth.Any())
            {
                EatSeedlingFromMouth();
            }
        }
    }

    private void EatSeedlingFromMouth()
    {
        Eat();
        _mouth.RemoveAt(0);
    }

    public void EatSeedlingFromTile()
    {
        Eat();
        _plane.HamsterIsEatingSeedlings(this);
    }

    private void Eat()
    {
        IsHungry = false;
        Representation = FedRepresentation;
    }

    public void StoreInMouth()
    {
        var seedling = _plane.GetSeedlingOn(Position);
        _mouth.Add(seedling);
        _plane.HamsterIsStoringSeedlings(this);
    }

    public override string ToString()
    {
        return Representation;
    }
}

File: JAVA/modul_2_object_orientation/src/lerneinheiten/L02KlassenUndMethoden/hamster/Hamster.java
package lerneinheiten.L02KlassenUndMethoden.hamster;

import java.util.ArrayList;
import java.util.Random;

public class Hamster {
    // Felder
    int _xPosition;
    int _yPosition;
    String _representation;
    final static String _hungryRepresentation = "🤬";
    final static String _fedRepresentation = "🐹";
    boolean _isHungry;
    ArrayList<Seedling> _mouth = new ArrayList<>();

    // Hat-Beziehungen
    Plane _plane;
    // -----------------------------
    // Methoden
    void nahrungsVerhalten() {
        Random random = new Random();

        // Zufällig hungrig werden
        if (random.nextDouble() < 0.1)
        {
            _isHungry = true;
            _representation = _hungryRepresentation;
        }

        boolean tileHasSeedling = _plane.tileTakenBySeedling(_xPosition, _yPosition);

        if (tileHasSeedling)
        {
            if (_isHungry)
            {
                eatSeedlingFromTile();
            }
            else
            {
                storeInMouth();
            }
        }
        else
        {
            if (_isHungry && !_mouth.isEmpty())
            {
                eatSeedlingFromMouth();
            }
        }
    }

    void eat() {
        _representation = _fedRepresentation;
        _isHungry = false;
    }

    void eatSeedlingFromMouth()
    {
        eat();
        _mouth.remove(0);
    }

    void eatSeedlingFromTile()
    {
        eat();
        _plane.hamsterIsEatingSeedlings(this);
    }

    void storeInMouth()
    {
        Seedling seedling = _plane.getSeedlingOn(_xPosition, _yPosition);
        _mouth.add(seedling);
        _plane.hamsterIsStoringSeedlings(this);
    }

    void bewegen() {
        Random random = new Random();
        int index = random.nextInt(0, Direction.values().length);
        Direction direction = Direction.values()[index];

        _plane.bewegeHamster(this, direction);
    }

    // Konstruktor
    Hamster(Plane plane) {
        boolean done;
        _plane = plane;
        _representation = _hungryRepresentation;
        _isHungry = true;
        int xWunsch;
        int yWunsch;

        // zufällig den Hamster positionieren.
        do {
            Random random = new Random();
            xWunsch = random.nextInt(0, _plane._size);
            yWunsch = random.nextInt(0, _plane._size);

            done = _plane.assignInitialPosition(this, xWunsch, yWunsch);
        } while (!done);

        _xPosition = xWunsch;
        _yPosition = yWunsch;
    }
}
